### PR TITLE
Per-file validation loss logging and --tracking_peak_path

### PR DESCRIPTION
**Pull request by PigeonMark**
**Summary**

Two related changes to the validation pipeline during training.

1. **Per-file validation loss**: each file passed via -p/--validation_peak_path now gets its own DataLoader and logs valid_CELoss/<stem> independently, in addition to the existing aggregate valid_CELoss that drives checkpoint selection. This makes it easy to spot per-dataset regression in TensorBoard/CSV logs without affecting the checkpoint criterion.
2. **Tracking-only validation files**: a new -t/--tracking_peak_path option accepts additional annotated files that are logged per-file (valid_CELoss/<stem>) but excluded from the aggregate valid_CELoss. The intended use case is monitoring catastrophic forgetting on a held-out dataset during fine-tuning, where including it in the aggregate would bias best-checkpoint selection.

**Changes**

- casanovo/denovo/dataloaders.py: DeNovoDataModule accepts tracking_paths; setup() builds one AnnotatedSpectrumDataset per validation/tracking file; val_dataloader() returns a list (main files first, tracking files after); exposes val_stems, n_main_loaders
- casanovo/denovo/model.py: validation_step gains a dataloader_idx parameter; logs valid_CELoss/<stem> for every loader and valid_CELoss only for main loaders (dataloader_idx < n_main_loaders); all self.log calls use add_dataloader_idx=False to prevent Lightning from appending /dataloader_idx_N suffixes that would break ModelCheckpoint(monitor="valid_CELoss"); on_validation_epoch_end records per-file keys in _history
- casanovo/denovo/model_runner.py: train() accepts tracking_peak_path; sets model.val_stems and model.n_main_loaders before trainer.fit(); initialize_data_module() forwards tracking_paths
- casanovo/casanovo.py: -t/--tracking_peak_path CLI option
- tests/unit_tests/test_unit.py: test_train_val_step_functions fixed (now passes batch_idx); 3 new tests: per-file dataset wiring (val_stems, n_main_loaders, loader count), validation_step log-key assertions with mocked self.log, CLI -t argument parsing
- tests/unit_tests/test_runner.py: 2 new runner tests: backward-compat (no tracking path), per-file key in _history when tracking path is supplied

## Summary by CodeRabbit

* **New Features**
  * Repeatable CLI option (-t / --tracking_peak_path) to monitor validation on extra files without affecting checkpoint selection.
  * Per-file validation loss logging (metrics recorded per-file using distinct stems) and support for separate "tracking-only" validation files.

* **Tests**
  * Added unit tests for tracking validation behavior, per-file logging, dataloader ordering, and CLI forwarding of the new option.

* **Documentation**
  * CHANGELOG updated with the new entries.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Noble-Lab/casanovo/pull/637?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by PigeonMark** [comment]
@coderabbitai review

**Comment by PigeonMark** [comment]
@coderabbitai review

File: casanovo/denovo/dataloaders.py [hunk 1]
```diff
@@ -157,17 +189,37 @@ def setup(self, stage: str = None, annotated: bool = True) -> None:
...
+                        [path],
+                        annotated=True,
+                        mode=f"tracking_{i}",
+                        shuffle=False,
+                    )
+                )
+            self.n_main_loaders = len(self.valid_datasets)
+            self.val_stems = _unique_stems(
+                (self.valid_paths or []) + (self.tracking_paths or [])
+            )
```
[the remark this answers is not in the document]
Comment (PigeonMark) [reply casanovo/denovo/dataloaders.py:235]: These findings are not valid.

The fallback `n_main = self.n_main_loaders if self.n_main_loaders > 0 else 1` exists for backward compatibility: when the model is loaded from a pre-PR checkpoint, `n_main_loaders` defaults to 0, and the one validation dataloader is the main one. The fallback ensures valid_CELoss is still logged.

The "tracking-only with no valid_paths" scenario (`n_main_loaders == 0` set explicitly) is not possible, the CLI defaults `validation_peak_path` to `train_peak_path` when empty and `ModelRunner.train()` requires `valid_peak_path`.

**Comment by PigeonMark** [comment]
All coderabbitai comments resolved, ready for review @bittremieux

File: casanovo/denovo/dataloaders.py [hunk 2]
```diff
@@ -24,6 +24,25 @@
 logger = logging.getLogger("casanovo")
 
 
+def _unique_stems(paths: list) -> list:
+    """Return stems from *paths*, disambiguating duplicates with a suffix."""
```
Comment (bittremieux) [inline casanovo/denovo/dataloaders.py]: **todo:** Please provide a complete docstring, including parameters and output information.

File: casanovo/denovo/dataloaders.py [hunk 3]
```diff
@@ -157,17 +189,37 @@ def setup(self, stage: str = None, annotated: bool = True) -> None:
...
-            if self.valid_paths is not None:
-                self.valid_dataset = self._make_dataset(
-                    self.valid_paths,
-                    annotated=True,
-                    mode="valid",
-                    shuffle=False,
+            # Build one dataset per validation file so each gets its own
+            # DataLoader and its loss can be logged separately.
+            self.valid_datasets = []
+            for i, path in enumerate(self.valid_paths or []):
```
Comment (bittremieux) [inline casanovo/denovo/dataloaders.py]: **nitpick:** `or []` seems superfluous with line 126, code can be simplified a bit.

File: casanovo/denovo/dataloaders.py [hunk 4]
```diff
@@ -157,17 +189,37 @@ def setup(self, stage: str = None, annotated: bool = True) -> None:
...
+                        annotated=True,
+                        mode=f"valid_{i}",
+                        shuffle=False,
+                    )
                 )
-                logger.info(
-                    "Validation dataset contains %d spectra.",
-                    self._get_n_spectra(self.valid_dataset),
+            self.tracking_datasets = []
+            for i, path in enumerate(self.tracking_paths or []):
```
Comment (bittremieux) [inline casanovo/denovo/dataloaders.py]: **nitpick:** Same here.

File: casanovo/denovo/dataloaders.py [hunk 5]
```diff
@@ -300,9 +352,12 @@ def train_dataloader(self) -> torch.utils.data.DataLoader:
         """Get the training DataLoader."""
         return self._make_loader(self.train_dataset, shuffle=self.shuffle)
 
-    def val_dataloader(self) -> torch.utils.data.DataLoader:
-        """Get the validation DataLoader."""
-        return self._make_loader(self.valid_dataset)
+    def val_dataloader(self) -> list:
+        """Get validation DataLoaders — one per file, main first then tracking."""
```
Comment (bittremieux) [inline casanovo/denovo/dataloaders.py]: **todo:** Update docstring.

File: casanovo/denovo/model.py [hunk 6]
```diff
@@ -837,33 +840,66 @@ def training_step(
...
+            to the aggregate ``valid_CELoss`` used for checkpoint selection.
+            Higher indices are "tracking" files logged per-file only.
 
         Returns
         -------
         torch.Tensor
             The loss of the validation step.
         """
-        # Record the loss.
-        loss = self.training_step(batch, mode="valid")
```
Comment (bittremieux) [inline casanovo/denovo/model.py:863]: **todo:** I guess if we can't reuse the `training_step` anymore, that function can be simplified.

**Review by bittremieux — CHANGES_REQUESTED** [review]

**Review by bittremieux — APPROVED** [review]
